provisioning.go

```diff
@@ -0,0 +1,370 @@
+		return xerrors.Errorf("cannot set the lockout hierarchy authorization value: %w", err)
+	}
+
+	return nil
+}
+
+// RequestTPMClearUsingPPI submits a request to the firmware to clear the TPM on the next reboot. This is the only way to clear
+// the TPM if owner clear has been disabled for the TPM, or the lockout hierarchy authorization value has been set previously but
+// is unknown.
+func RequestTPMClearUsingPPI() error {
+	f, err := os.OpenFile(ppiPath, os.O_WRONLY, 0)
+	if err != nil {
+		return xerrors.Errorf("cannot open request handle: %w", err)
+	}
+	defer f.Close()
+
+	if _, err := f.WriteString(clearPPIRequest); err != nil {
+		return xerrors.Errorf("cannot submit request: %w", err)
+	}
+
+	return nil
+}
+
+// ProvisionStatus returns the provisioning status for the specified TPM.
+func ProvisionStatus(tpm *TPMConnection) (ProvisionStatusAttributes, error) {
+	var out ProvisionStatusAttributes
+
+	session := tpm.HmacSession().IncludeAttrs(tpm2.AttrAudit)
+
+	if ek, err := tpm.CreateResourceContextFromTPM(ekHandle, session); err != nil {
+		if _, unavail := err.(tpm2.ResourceUnavailableError); !unavail {
+			return 0, err
+		}
+	} else if ekInit, err := tpm.EndorsementKey(); err == nil && bytes.Equal(ekInit.Name(), ek.Name()) {
+		out |= AttrValidEK
+	}
+
+	if srk, err := tpm.CreateResourceContextFromTPM(srkHandle, session); err != nil {
+		if _, unavail := err.(tpm2.ResourceUnavailableError); !unavail {
+			return 0, err
+		}
+	} else if tpm.provisionedSrk != nil {
+		if bytes.Equal(tpm.provisionedSrk.Name(), srk.Name()) {
+			out |= AttrValidSRK
+		}
+	} else if ok, err := isObjectPrimaryKeyWithTemplate(tpm.TPMContext, tpm.OwnerHandleContext(), srk, &srkTemplate, tpm.HmacSession()); err != nil {
+		return 0, xerrors.Errorf("cannot determine if object at 0x%08x is a primary key in the storage hierarchy: %w", srkHandle, err)
+	} else if ok {
+		out |= AttrValidSRK
+	}
+
+	props, err := tpm.GetCapabilityTPMProperties(tpm2.PropertyMaxAuthFail, 3)
+	if err != nil {
+		return 0, xerrors.Errorf("cannot fetch DA parameters: %w", err)
+	}
+	if props[0].Value <= maxTries && props[1].Value >= recoveryTime && props[2].Value >= lockoutRecovery {
+		out |= AttrDAParamsOK
+	}
+
+	props, err = tpm.GetCapabilityTPMProperties(tpm2.PropertyPermanent, 1)
+	if err != nil {
+		return 0, xerrors.Errorf("cannot fetch permanent properties: %w", err)
+	}
+	if tpm2.PermanentAttributes(props[0].Value)&tpm2.AttrDisableClear > 0 {
+		out |= AttrOwnerClearDisabled
+	}
+	if tpm2.PermanentAttributes(props[0].Value)&tpm2.AttrLockoutAuthSet > 0 {
+		out |= AttrLockoutAuthSet
+	}
+
+	if lockIndex, err := tpm.CreateResourceContextFromTPM(lockNVHandle, session); err != nil {
+		if _, unavail := err.(tpm2.ResourceUnavailableError); !unavail {
+			return 0, err
+		}
+	} else if _, err := readAndValidateLockNVIndexPublic(tpm.TPMContext, lockIndex, session); err == nil {
+		out |= AttrLockNVIndex
+	}
+
+	return out, nil
```

Comment:
The logic in this block is a bit hard to follow, perhaps we could refactor it a little bit to make it easier for the casual reader to understand what's going on? Maybe something like the expanded version below could be more readable.
```
        if lockIndex, err := tpm.CreateResourceContextFromTPM(lockNVHandle, session); err != nil {
                if _, unavail := err.(tpm2.ResourceUnavailableError); !unavail {
                        return 0, err
                } else {
                        return out, nil
                }       
        }

        if _, err := readAndValidateLockNVIndexPublic(tpm.TPMContext, lockIndex, session); err != nil {
                return out, nil
        }

        return out | AttrLockNVIndex, nil
```